src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -495,40 +495,193 @@ void Lowering::ContainBlockStoreAddress(GenTreeBlk* blkNode, unsigned size, GenT
+        {
+            lowerBound = comp->gtNewSimdZeroNode(TYP_SIMD16, simdBaseJitType, 16, /* isSimdAsHWIntrinsic */ false);
+            BlockRange().InsertAfter(castOp, lowerBound);
+            LowerNode(lowerBound);
+        }
+        else
+        {
+            GenTree* lowerBoundCns;
+
+            if (castToType == TYP_BYTE)
+            {
+                lowerBoundCns = comp->gtNewDconNode(-128.0, castFromType);
+            }
+            else
+            {
+                assert(castToType == TYP_SHORT);
+                lowerBoundCns = comp->gtNewDconNode(-32768.0, castFromType);
+            }
+
+            BlockRange().InsertAfter(castOp, lowerBoundCns);
+
+            lowerBound = comp->gtNewSimdHWIntrinsicNode(TYP_SIMD16, lowerBoundCns, NI_Vector128_CreateScalarUnsafe,
+                                                        simdBaseJitType, 16);
+            BlockRange().InsertAfter(lowerBoundCns, lowerBound);
+            LowerNode(lowerBound);
+        }
+
+        GenTree* simdMax =
+            comp->gtNewSimdHWIntrinsicNode(TYP_SIMD16, castOp, lowerBound, simdMaxIntrinId, simdBaseJitType, 16);
+        BlockRange().InsertAfter(lowerBound, simdMax);
+        LowerNode(simdMax);
+
+        // We min ourselves against the upper boun
+        GenTree* upperBound;
+        GenTree* upperBoundCns;
+
+        switch (castToType)
+        {
+            case TYP_BYTE:
+            {
+                upperBoundCns = comp->gtNewDconNode(+127.0, castFromType);
+                break;
+            }
+
+            case TYP_UBYTE:
+            {
+                upperBoundCns = comp->gtNewDconNode(+255.0, castFromType);
+                break;
+            }
+
+            case TYP_SHORT:
+            {
+                upperBoundCns = comp->gtNewDconNode(+32767.0, castFromType);
+                break;
+            }
+
+            case TYP_USHORT:
+            {
+                upperBoundCns = comp->gtNewDconNode(+65535.0, castFromType);
+                break;
+            }
+
+            default:
+            {
+                unreached();
+            }
+        }
+
+        BlockRange().InsertAfter(simdMax, upperBoundCns);
+
+        upperBound = comp->gtNewSimdHWIntrinsicNode(TYP_SIMD16, upperBoundCns, NI_Vector128_CreateScalarUnsafe,
+                                                    simdBaseJitType, 16);
+        BlockRange().InsertAfter(upperBoundCns, upperBound);
+        LowerNode(upperBound);
+
+        GenTree* simdMin =
+            comp->gtNewSimdHWIntrinsicNode(TYP_SIMD16, simdMax, upperBound, simdMinIntrinId, simdBaseJitType, 16);
+        BlockRange().InsertAfter(upperBound, simdMin);
+        LowerNode(simdMin);
+
+        // Replace the castOp with the saturated value
+        castOp = comp->gtNewSimdHWIntrinsicNode(castFromType, simdMin, NI_Vector128_ToScalar, simdBaseJitType, 16);
+        BlockRange().InsertAfter(simdMin, castOp);
+        LowerNode(castOp);
+
+        tmpType = TYP_INT;
```

Comment:
This shouldn't generate a `int -> int` cast. This path is only hit when `varTypeIsSmall(castToType)`.

The `tmpType` needs to be `int` because the underlying instruction produces a `TYP_INT`.

Reply:
Hmm, so as I am reading the code, here's what this method does:
1) We have `tree` as `CAST(small type <- float)`.
2) We change it to `CAST(small type <- float (saturation))`.
3) We change it then to `CAST(small type <- CAST(int <- float (saturation)))`.

So my question is: why do we need the intermediate cast? Presumably we have ensured that the values here are in correct ranges with the saturation.

(Sorry for the confusion here, `int -> int` because in terms of values `CAST(small type <- float (saturation))` produces `int`s)

Reply:
I believe the answer is we shouldn't need it and we merely need to change from `CAST(small type <- float)` to `CAST(TYP_INT <- float (saturation))`

The value, due to the saturation, will be in bounds, but we need a `TYP_INT` node here.